SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp

```diff
@@ -576,14 +577,28 @@ void USpatialNetDriver::OnGSMQuerySuccess()
 
 void USpatialNetDriver::RetryQueryGSM()
 {
+	float WatchdogTimer = 10.0;
 	float RetryTimerDelay = SpatialConstants::ENTITY_QUERY_RETRY_WAIT_SECONDS;
 
+	if (!GSMQueryWatchdog.IsValid())
+	{
+		TimerManager.SetTimer(
+			GSMQueryWatchdog,
+			[WatchdogTimer]() {
+				UE_LOG(LogSpatialOSNetDriver, Log,
+					   TEXT("Worker failed to startup after %f seconds. Usually caused by servers not registering themselves with the "
+							"deployment yet. Did you launch the correct number of servers?"),
+					   WatchdogTimer);
+			},
+			WatchdogTimer, false);
+	}
```

Comment:
I'm not sure we need an extra timer manager to control log rate. Could we have a float timeout for when to print logs and pass an float for the time since we last printed a log into the timer below?

Reply:
Also, can the log be something more like "Client waiting for SpatialOS servers to be ready" and then ideally appending the actual reason, that we figure out in the query delegate function, e.g. `!bQueryResponseSuccess` or `!bNewAcceptingPlayers`

Reply:
So proper rate limiting of all the relevant startup logs will be difficult, because they are spread over several classes, and some phases are stateless (like the GSM query).
We could imagine some utility macro wrapping some relevant logs, and keep them around to be queried when the time has come, but it's starting to look a bit overkill.